webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/write/PolicyWriterService.scala

```diff
@@ -789,75 +862,91 @@ class PolicyWriterServiceImpl(
                               val replace = (s: String) => StringUtils.replace(s, Policy.TAG_OF_RUDDER_MULTI_POLICY, id.getRudderUniqueId)
                               (replace(filled), replace(templateInfo.destination))
                           }
-      _                <- tryo { FileUtils.writeStringToFile(new File(outPath, dest), replaced, StandardCharsets.UTF_8) } ?~!
-                            s"Bad format in Technique ${templateInfo.id.toString} (file: ${templateInfo.destination})"
+      _                <- File(outPath, dest).createParentsAndWrite(replaced).chainError(
+                              s"Bad format in Technique ${templateInfo.id.toString} (file: ${templateInfo.destination})"
+                          )
+      t2               <- currentTimeNanos
+      _                <- writeTimer.writeTemplate.update(_ + t2 - t1)
     } yield {
       outPath
     }
   }
 
   /**
    * Move the machine promises folder  to the backup folder
-   * @param machineFolder
+   * @param nodeFolder
    * @param backupFolder
    */
-  private[this] def backupNodeFolder(nodeFolder: String, backupFolder: String): Unit = {
-    val src = new File(nodeFolder)
-    if (src.isDirectory()) {
-      val dest = new File(backupFolder)
-      if (dest.isDirectory) {
-        // force deletion of previous backuup
-        FileUtils.deleteDirectory(dest)
+  private[this] def backupNodeFolder(nodeFolder: String, backupFolder: String): IOResult[Unit] = {
+    IOResult.effect {
+      val src = File(nodeFolder)
+      if (src.isDirectory()) {
+        val dest = File(backupFolder)
+        if (dest.isDirectory) {
```

Comment:
just a stupid question: if the folder is there, and we move, do we really nned to delete first ?

Reply:
I think it was for the case when something bad happened in a previous generation. But now that we can be much more fine grained on our error management, perhaps we could delete on error. I'm not sur it changes much in perf (to avoid the unused delete).

Reply:
delete is fairly costly when there are many files and already a lors of I/O
I don't remember exactly, but it was between 2 and 6 minutes to delete with rm the backup folders with 10 000 nodes